### DISCUSSION TITLE
关于表单模型的生命周期

### DISCUSSION THREAD

**Discussion by JustDs** [alibaba/formily#1606] [Ideas]
目前 Formily 的表单包括下面几个生命周期节点：

- `Init`：在 Form 初始化，也即 `createForm()` 的时候触发
- `Mount`：在表单对应的视图层组件挂载时，通过 `useAttach()` 触发
- `Unmount`：在表单对应的视图层组件卸载时，通过 `useAttach()` 触发

而 Formily 在每个生命周期节点，除了触发对应节点生命周期事件以外，做的事情包括：

- `Init`：表单状态初始化、包装为 `Observable`、observe 表单值以触发后续生命周期事件
https://github.com/alibaba/formily/blob/bb4f175f6bbda1201439032bdecce1d044a5a726/packages/core/src/models/Form.ts#L85-L91
https://github.com/alibaba/formily/blob/bb4f175f6bbda1201439032bdecce1d044a5a726/packages/core/src/models/Form.ts#L170-L183
https://github.com/alibaba/formily/blob/bb4f175f6bbda1201439032bdecce1d044a5a726/packages/core/src/models/Form.ts#L590-L593

- `Mount`：设置表单状态
https://github.com/alibaba/formily/blob/bb4f175f6bbda1201439032bdecce1d044a5a726/packages/core/src/models/Form.ts#L595-L601

- `Unmount`：设置表单状态、释放所有的内部缓存、释放所有的 observe 和 subscribe
https://github.com/alibaba/formily/blob/bb4f175f6bbda1201439032bdecce1d044a5a726/packages/core/src/models/Form.ts#L603-L614

可以看到，初始化方面的事情都在 `Init` 中，`Mount` 中几乎没有做任何实际的事情，而资源的释放都是在 `Unmount` 中的。

这样一来，Form 在拥有视图层的情况下就变成了只能一次性使用的东西。无论是有意（例如将表单放在弹窗 / 抽屉里）还是无意（例如某些问题导致意外重渲染）地卸载重挂了表单组件，就会导致 Form 模型被销毁。而如果 Form 的组件没有在卸载重挂的过程中同步重新创建（例如在组件外单例创建，或是在上层组件创建），会让重新挂载的组件引用到已被销毁的 Form 模型，从而产生一些莫名的 bug。这一点甚至没有任何报错提示，也没有文档指引用户注意这一点。

个人认为更合理的做法是增加 `Destroy` 节点，让 `Destroy` 对偶 `Init`、`Unmount` 对偶 `Mount`。原先放在 `Unmount` 中释放资源的逻辑改为由 `Destroy` 负责，`Unmount` 节点变为与 `Mount` 类似只记录状态。

在这样调整之后，用户如果在某个 scope 下通过 `createForm()` 手动创建了 Form 模型，在离开 scope 时也需要手动通过 `form.destroy()` 来进行销毁以免内存泄漏。考虑到历史版本的兼容性，可以选择让 `useAttach()` 是在组件卸载的时候只触发 `Unmount` 还是先后触发 `Unmount` 和 `Destroy`。

本问题的上下文参见：#1602，可以在 discussion 这边讨论下更合适的做法，如果有空的话我晚点会提个 PR 看看可行性。

**Comment by janryWang** [comment]
先详细描述一下你场景的需求，和问题/痛点？

**Reply by JustDs** [reply]
场景：一个在可折叠抽屉内的配置表单，配置出的东西需要在抽屉外实时预览
问题：抽屉折叠再打开之后，`onFieldValueChange` 之类的钩子失效了

**Reply by janryWang** [reply]
是不是和 https://designable-antd.formilyjs.org/ 这上面的折叠表单相似？

**Reply by JustDs** [reply]
是 稍微有些类似

**Reply by JustDs** [reply]
具体场景的话，我其实已经找产品从需求角度绕开了，但是生命周期的问题我依然觉得是个应当解决的问题。

**Reply by janryWang** [reply]
> 是 稍微有些类似

最好能给个交互稿看看吧。为啥会有这种场景，因为formily表单设计器也是足够复杂的产品了，也没遇到这种问题

**Reply by JustDs** [reply]
毕竟是还没上线的商业项目哈，交互稿可能不太方便公开，也没空自己简化一遍再发出来。

我想表达的意思是：**我遇到的** 具体场景问题已经不是重点了，毕竟问题已经绕开了也就不再需要解决了；但是 Form 的 create 和 attatch 没放在一起的情况是一定可能出现的，因为没有任地方限制这么做，也没有任何指引要求使用者不要这么做，出现这种情况再加上 Form 的组件被卸载重挂就会出现这个问题，而真出现问题了还会很隐蔽难以排查。

我现在诉求并不是解决我碰到的个例问题，我只是觉得相比要求使用者不要以设计者预计以外的方式来使用，把库的逻辑做的更加自洽、更加符合直觉是更好的做法，所以发起这个 discussion 的本意在于讨论方案的合理性。

**Reply by janryWang** [reply]
明白你说的意思，我也更加注重自洽性，但是，如果抛开场景谈设计，都有点像耍流氓，如果实在没有那种真的让人很痛的场景，我宁可先保持不变，因为新增API或者break change，对大家的成本都会提升很多

**Comment by liuweiGL** [comment]
按我的想法是 createForm 应该换成 useCreateForm/useNewForm/useFormInstance 这种 hook 更直接？

**Reply by JustDs** [reply]
我觉得这是一个可行的解决方案，通过 hooks 创建的 Form 也负责在退出 scope 时，而通过 createForm 手动创建的 Form 也手动销毁。

**Comment by liuweiGL** [comment]
核心问题就是 createForm 可以当成单例模式来使用（特别是刚接触 formily 的用户，感觉这样可以提升性能），而忽略了创建出来的 form 实例是一次性的，最终可能会导致难以排查的 bug，不知道理解对了吗？

**Reply by JustDs** [reply]
是的，这个说法比较接近我想表达的意思。

**Comment by janryWang** [comment]
我的核心观点：
1. FormProvider的定位，是用于下发上下文，同时自动维护form实例的挂载和卸载事件，这样用户可以无需关注内存回收的问题
2. 要看具体场景，是哪种用法更多，API设计才会偏向哪种用户，对于第一点，我个人觉得是属于普适性需求，肯定要照顾到大多数用户
3. 最好能把你的具体场景描述清楚，方便分析到底是使用上的反模式，还是的确库本身设计的有问题才会导致出现这样痛的痛点，现在我主要是没感受到你的痛点，给个体感更强的案例的话，也可以帮助我从实际场景上判断设计的好坏

**Reply by JustDs** [reply]
好的，如果后续还有具体场景，我会继续再这里补充。

我这次遇到的问题，确实可以说是使用上的反模式，或者说是产品需求上的反模式，所以才调整了需求绕开了。

对于默认销毁这件事我是不反对的，这样确实更方便，我只是希望能够提供不销毁的选项。

正是因为 breaking change 对使用者而言成本会很高，所以我才更觉得应该及早在 beta 版本把生命周期这些比较重要的概念整理清楚，以免给以后留下很难修补的坑。

**Reply by janryWang** [reply]
> 正是因为 breaking change 对使用者而言成本会很高，所以我才更觉得应该及早在 beta 版本把生命周期这些比较重要的概念整理清楚，以免给以后留下很难修补的坑。

非常赞同你的说法，所以我也比较迫切想知道目前的设计到底问题在哪里，得把本质问题梳理清楚才方便设计，而不是遇到问题就立马改，也没想清楚它到底是不是最优的设计，所以还是需要权衡一下的

**Reply by JustDs** [reply]
今天正好遇到了一个与此话题有关的新案例，做了个最小复现的 demo：

[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/small-water-g9ypj?file=/App.tsx)

我知道这里面的实现在你看来可能是反范式的，可否给出一个你觉得符合范式的修改方案？

**Reply by janryWang** [reply]
https://codesandbox.io/s/cool-lake-05jx4 
合理的方案是FormProvider和createForm是在同一个组件内搭配使用，且是一对一的搭配，而不是FormProvider散落在各处，form只有一个实例，这个给用户的感觉就是formily可以支持多个FormProvider共享同一个form实例了，显然，这会放大问题域了，可能就会有一大批用户乱用这种模式，虽然我不知道这种模式到底好不好，因为也没有具体能体现出这种模式价值的点，所以就是感觉怪怪的，不可预知它的好坏是有点恐怖的

**Reply by liuweiGL** [reply]
> 今天正好遇到了一个与此话题有关的新案例，做了个最小复现的 demo：
> 
> [![Edit on CodeSandbox](https://camo.githubusercontent.com/90808661433696bc57dce8d4ad732307b5cec6270e6b846f114dcd7ee7f9458a/68747470733a2f2f636f646573616e64626f782e696f2f7374617469632f696d672f706c61792d636f646573616e64626f782e737667)](https://codesandbox.io/s/small-water-g9ypj?file=/App.tsx)
> 
> 我知道这里面的实现在你看来可能是反范式的，可否给出一个你觉得符合范式的修改方案？

这个 demo 想表达什么问题呢，没 get 到。

**Reply by JustDs** [reply]
> > 今天正好遇到了一个与此话题有关的新案例，做了个最小复现的 demo：
> > [![Edit on CodeSandbox](https://camo.githubusercontent.com/90808661433696bc57dce8d4ad732307b5cec6270e6b846f114dcd7ee7f9458a/68747470733a2f2f636f646573616e64626f782e696f2f7374617469632f696d672f706c61792d636f646573616e64626f782e737667)](https://codesandbox.io/s/small-water-g9ypj?file=/App.tsx)
> > 我知道这里面的实现在你看来可能是反范式的，可否给出一个你觉得符合范式的修改方案？
> 
> 这个 demo 想表达什么问题呢，没 get 到。

点在于：这个 demo 下，实际上第一次关闭弹窗表单模型就已经被销毁了，如果我往上挂点 effects，从第二次打开弹窗开始就不会触发了，甚至还还会有其他我不知道的 bug。

问题更大的点在于：作为项目的维护者你都没有感受到这是个明显错误的使用方式，会造成 bug，普通使用者又从何感知到这点呢？

**Reply by JustDs** [reply]
> https://codesandbox.io/s/cool-lake-05jx4
> 合理的方案是FormProvider和createForm是在同一个组件内搭配使用，且是一对一的搭配，而不是FormProvider散落在各处，form只有一个实例，这个给用户的感觉就是formily可以支持多个FormProvider共享同一个form实例了，显然，这会放大问题域了，可能就会有一大批用户乱用这种模式，虽然我不知道这种模式到底好不好，因为也没有具体能体现出这种模式价值的点，所以就是感觉怪怪的，不可预知它的好坏是有点恐怖的

这个修改方式我认为是很合理的，`FormProvider` 与 `createForm()` 一对一搭配使用我觉得可能是很重要的、并且未被官方明确表达出来的点。是否可以考虑基于这个做做封装试试效果？

**Comment by liuweiGL** [comment]
我觉得对外的话还是应该提供一个 hook，至少在 react 中应该是需要的，不然每个用户都得自己封装一遍。

**Reply by janryWang** [reply]
不好提供hook，真要提供，也是useMemo包一下，反而增加理解成本了

**Comment by javahuang** [comment]
刚刚遇到了这个问题，竟然在论坛上也翻到了，真巧。
正在使用 formily2 重构问卷设计器，我的每个问题、选项设置都分别对应一个 Form。从一个设置 Drawer 切换到另外一个设置 Drawer 时，我想恢复某些 Field 的 caches.value，但是由于 Form unmount 的时候调用了 `onUnmount()` 导致 fields 状态都丢失了。
我现在是直接使用 <FormContext.Provider>，然后在需要销毁 form 的时候手动的执行 `onUnmount()`。

**Reply by janryWang** [reply]
你这个场景太特殊了，不同的表单之间切换，还要把内置状态给同步过去，不是一个很好的交互行为，假如相同字段名上用的不同的组件，但是组件接收和输出的值类型都不一样，这样切换过去是会出问题的
